src/stream.rs

```diff
@@ -1,101 +1,205 @@
+    ) -> OutputStreamBuilder {
+        self.config = OutputStreamConfig {
+            channel_count: config.channels(),
+            sample_rate: config.sample_rate(),
+            // In case of supported range limit buffer size to avoid unexpectedly long playback delays.
+            buffer_size: clamp_supported_buffer_size(config.buffer_size(), 1024),
+            sample_format: config.sample_format(),
+            ..self.config
+        };
+        self
+    }
 
-            devices
-                .find_map(|d| Self::try_from_device(&d).ok())
-                .ok_or(original_err)
+    pub fn with_config(mut self, config: &cpal::StreamConfig) -> OutputStreamBuilder {
+        self.config = OutputStreamConfig {
+            channel_count: config.channels,
+            sample_rate: config.sample_rate,
+            buffer_size: config.buffer_size,
+            ..self.config
+        };
+        self
+    }
+
+    pub fn open_stream(&self) -> Result<OutputStream, StreamError> {
+        let device = self.device.as_ref().expect("output device specified");
+        OutputStream::open(device, &self.config)
+    }
+
+    /// Try to open a new output stream with the builder's current stream configuration.
+    /// Failing that attempt to open stream with other available configurations
+    /// provided by the device.
+    /// If all attempts to open stream have not succeeded returns initial error.
+    pub fn try_open_stream(&self) -> Result<OutputStream, StreamError> {
+        let device = self.device.as_ref().expect("output device specified");
+        OutputStream::open(device, &self.config).or_else(|err| {
+            for supported_config in supported_output_configs(device)? {
+                if let Ok(handle) = Self::default()
+                    .with_supported_config(&supported_config)
+                    .open_stream()
+                {
+                    return Ok(handle);
+                }
+            }
+            Err(err)
         })
     }
+
+    /// Try to open a new output stream for the default output device with its default configuration.
+    /// Failing that attempt to open output stream with alternative configuration and/or non default
+    /// output devices. Returns stream for first tried configuration that succeeds.
+    /// If all attempts have not succeeded return the initial error.
+    pub fn try_default_stream() -> Result<OutputStream, StreamError> {
+        Self::from_default_device()
+            .and_then(|x| x.open_stream())
+            .or_else(|original_err| {
+                let mut devices = match cpal::default_host().output_devices() {
+                    Ok(devices) => devices,
+                    Err(err) => {
+                        #[cfg(feature = "tracing")]
+                        tracing::error!("error getting list of output devices: {err}");
+                        #[cfg(not(feature = "tracing"))]
+                        eprintln!("error getting list of output devices: {err}");
```

Comment:
I do not like this pattern. Since one will have to repeat this pair of conditional blocs anywhere a log message is required. Tracing lib is fast but requires to keep some data in memory and creating spans. So it should not be enabled by default. It is best suited for performance related stuff and debugging, but here an error is reported which better be available by default. I'd use a `log` crate in such cases or maybe have a custom macro that can be used within `rodio`. Alternatively one can return composite errors that will contain all possible information and let users to decide what to do with it (using e.g. `error-stack` crate).

Reply:
> I'd use a log crate in such cases

tracing supports log, but log not tracing which is why we went with tracing. 

>Tracing lib is fast but requires to keep some data in memory and creating spans. So it should not be enabled by default.

Spans are not created in rodio so its not an issue. Given the print/error happens on an error path this should never be a perf issue. Unless we get data showing otherwise I personally do not see a perf problem. 

> So it should not be enabled by default.

Printing disturbs tui clients which is why that is not the default.

> I do not like this pattern. Since one will have to repeat this pair of conditional blocks anywhere a log message is required. 

Agreed its not a nice pattern however,  right now we trace/eprint only in one or two places. Maybe in the future rodio could have a macro itself that calls either error or eprintln depending on the feature enabled or some other abstraction. But as it stands the abstraction is more hassle then the prints/trace calls. 

> Alternatively one can return composite errors that will contain all possible information and let users to decide what to do with it.

This pattern happens in a few other places where its in a callback so that is not possible. Here you are absolutely right, this should be part of an error enum.

>  (using e.g. error-stack crate).

We recently removed thiserror due to the compile time impact and as we hardly used it. Might re-evaluate that later but for now I say we just hand roll the enum.

Reply:
It is what most crates do however the `eprintln` was already here (it predates tracing & log I think). I do not know if its a breaking change to remove it. We can do that. An alternative is to let users pass in a callback. The callback would then get an error enum. Then the user can decide what happens.